sys/include/net/gcoap/dns.h

```diff
@@ -0,0 +1,200 @@
+#define CONFIG_GCOAP_DNS_SERVER_URI_LEN     64U
+#endif
+
+/**
+ * @brief   maximum number of credentials that can be added with @ref gcoap_dns_cred_add()
+ */
+#ifndef CONFIG_GCOAP_DNS_CREDS_MAX
+#if IS_USED(MODULE_GCOAP_DTLS) || defined(DOXYGEN)
+#define CONFIG_GCOAP_DNS_CREDS_MAX          1U
+#else
+#define CONFIG_GCOAP_DNS_CREDS_MAX          0U
+#endif
+#endif
+
+/**
+ * @brief   size of the buffer used to build a CoAP request
+ */
+#ifndef CONFIG_GCOAP_DNS_PDU_BUF_SIZE
+#define CONFIG_GCOAP_DNS_PDU_BUF_SIZE       128U
+#endif
+
+/**
+ * @brief   block-size used for requests (both Block2 control and Block1)
+ *
+ * Should be less than @ref CONFIG_GCOAP_DNS_PDU_BUF_SIZE or
+ * @ref CONFIG_GCOAP_PDU_BUF_SIZE and must be a power
+ * of 2
+ */
+#ifndef CONFIG_GCOAP_DNS_BLOCK_SIZE
+#define CONFIG_GCOAP_DNS_BLOCK_SIZE         64U
+#endif
+/** @} */
+
+/**
+ * @brief   Query a domain name via CoAP synchronously
+ *
+ * @note    Only available with module `gcoap_dns_async` compiled in.
+ *
+ * @param[in] domain_name   A '\0'-terminated domain name. Must not be NULL.
+ * @param[out] addr_out     The resolved address. Must not be NULL.
+ * @param[in] family        The desired address family for @p addr_out.
+ *                          @ref AF_UNSPEC for any address family (an IPv6
+ *                          address will take preference over an IPv4 address).
+ *
+ * @return  length of @p addr_out in bytes on success
+ * @return  -EAFNOSUPPORT, if the hostname of the URI resolves to an unknown address family.
+ * @return  -EBADMSG, when receiving erroneous response or response containing
+ *          an error code.
+ * @return  -ECONNABORTED, if CoAP request cannot be sent.
+ * @return  -ECONNREFUSED, if no URI is set for the client (see @ref gcoap_dns_server_uri_set()).
+ * @return  -EDESTADDRREQ, if CoAP response was received from an unexpected
+ *          remote.
+ * @return  -EHOSTUNREACH, if the hostname of the URI can not be resolved
+ * @return  -ENOBUFS, if there was not enough buffer space for the request.
+ * @return  -ENOBUFS, if length of received CoAP body is greater than
+ *          @ref CONFIG_DNS_MSG_LEN.
+ * @return  -ENOENT, if Zone-ID of the URI can not be found locally.
+ * @return  -ENOMSG, if CoAP response did not contain a DNS response.
+ * @return  -ENOTRECOVERABLE, on gCoAP-internal error.
+ * @return  -ENOTSUP, if credential can not be added for to client.
+ * @return  -ENOTSUP, if module `gcoap_dns_async` is not compiled in.
+ * @return  -ETIMEDOUT, if CoAP request timed out.
+ */
+int gcoap_dns_query(const char *domain_name, void *addr_out, int family);
+
+/**
+ * @brief   Sets and checks a URI
+ *
+ * @param[in] uri   A URI. May be NULL to remove the URI.
+ *
+ * @return  length of @p uri on success (0 when URI is removed).
+ * @return  -EINVAL, if @p uri is not a valid URI for DNS over CoAP.
+ * @return  -ENOBUFS, if the client implementation has not enough buffer space
+ *          to process the URI.
+ */
```

Comment:
Given that `gcoap_dns_server_uri_get` has a thread safety warning on it, I suppose this needs one too.